File: boj/class4/9663_backtracking.py
import sys

sys.stdin = open("input.txt", "r")
input = sys.stdin.readline

n = int(input())
ans = 0
queenX = [0 for i in range(n)]


def isAval(y, x):
    for i in range(y):
        if queenX[i] == x or abs(queenX[i] - x) == y - i:
            return False
    return True


def dfs(depth):
    for x in range(n):
        if isAval(depth, x):
            if depth == n - 1:
                global ans
                ans += 1
            else:
                queenX[depth] = x
                dfs(depth + 1)


dfs(0)
print(ans)

File: boj/class4/9663_backtracking.js
const fs = require("fs");
const inputPath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";

let n = fs.readFileSync(inputPath).toString().trim();
n = Number(n);
let ans = 0;
let queenX = new Array(n);

const isAval = (y, x) => {
  for (let i = 0; i < y; i++){
    if (queenX[i] === x || Math.abs(queenX[i] - x) === y - i) {
      return false;
    }
  }
  return true;
}

const dfs = (depth) => {
  for (let x = 0; x < n; x++) {
    if (isAval(depth, x)) {
      if (depth === n - 1) ans++;
      else {
        queenX[depth] = x;
        dfs(depth + 1);
      }
    }
  }
};

dfs(0);
console.log(ans);